### PR TITLE
Unify CombinationRepository and CombinationMultiShopRepository into one

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Unifies CombinationRepository and CombinationMultishopRepository. It is a part of code cleaning (will have to do the same with product repository). We introduced CombinationMultishopRepository to handle multishop cases and had in mind to later unify these repositories, so at the end only one should be left.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | 
| Related PRs       | 
| How to test?      | 
| Possible impacts? | Suppliers association needs to be checked. It has no multishop association, so I used default shop to get combinations where needed, need to make sure that works. Also SpecificPrices were impacted abit, need to check that too. Specific prices may need additional dynamic choice implementation (between shopId and combinationId), will document it later when its clear.

### CODE REVIEW

File: src/Adapter/Form/ChoiceProvider/CombinationIdChoiceProvider.php [hunk 1]
```diff
@@ -83,7 +84,10 @@ public function __construct(
     public function getChoices(array $options): array
     {
         $options = $this->resolveOptions($options);
-        $combinationIds = $this->combinationRepository->getCombinationIds(new ProductId($options['product_id']));
+        $combinationIds = $this->combinationRepository->getCombinationIds(
+            new ProductId($options['product_id']),
+            $options['shop_id'] ? ShopConstraint::shop($options['shop_id']) : ShopConstraint::allShops()
```
Comment (zuk3975) [inline src/Adapter/Form/ChoiceProvider/CombinationIdChoiceProvider.php:89]: Probably this combinationIdChoices and shop_id choices must dynamically update depending on each other.
Because in All shops context I should be able to apply spec.price for any combination in any shop and those combinations might be different in shops. So contextShop combinations != all shops combinations. So I think I should handle this in a dedicated PR via javascript :thinking: 
wdyt @jolelievre?

File: src/Core/Form/IdentifiableObject/DataProvider/SpecificPriceFormDataProvider.php [hunk 2]
```diff
@@ -67,6 +75,9 @@ public function getData($id): array
                 'currency_id' => $specificPriceForEditing->getCurrencyId(),
                 'country_id' => $specificPriceForEditing->getCountryId(),
                 'group_id' => $specificPriceForEditing->getGroupId(),
+                //@todo: this is also used to get combination choices based on shop, but probably it should be dynamic -
+                //       if all shops are chosen, then combination from all shops should be loaded to choice list,
+                //       if one shop is selected, then that shop combinations list should be loaded.
```
Comment (zuk3975) [inline src/Core/Form/IdentifiableObject/DataProvider/SpecificPriceFormDataProvider.php:80]: Probably combinationIdChoices and shopId choices must dynamically update depending on each other.
Because in All shops context I should be able to apply spec.price for any combination in any shop and those combinations might be different in shops. So contextShop combinations != all shops combinations. So I think I should handle this in a dedicated PR via javascript :thinking: 
wdyt @jolelievre?

File: src/Adapter/Form/ChoiceProvider/CombinationIdChoiceProvider.php (hunk 1 above)
Comment (jolelievre) [inline src/Adapter/Form/ChoiceProvider/CombinationIdChoiceProvider.php:89]: For this to work, option `shop_id` shouldn't be required but optional with default value null (which would be equivalent to All shops) This should maybe be moved into the PR that will handle the specific price form improvement

File: src/Adapter/Product/Update/ProductSupplierUpdater.php [hunk 3]
```diff
@@ -131,7 +140,10 @@ public function associateSuppliers(ProductId $productId, array $supplierIds): ar
         // We should always create an association not related to a combination
         $combinationIds = [new NoCombinationId()];
         if ($productType->getValue() === ProductType::TYPE_COMBINATIONS) {
-            $combinationIds = array_merge($combinationIds, $this->combinationRepository->getCombinationIds($productId));
+            $combinationIds = array_merge(
```
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php:143]: As we discussed earlier this one worries me a bit, I'm gonna work on suppliers multishop so I think it's preferable to wait to finish this PR until it's done This way we'll have tests to ensure the behaviour in multishop and we won't risk breaking some things with this refacto

File: src/Core/Form/IdentifiableObject/DataProvider/SpecificPriceFormDataProvider.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/DataProvider/SpecificPriceFormDataProvider.php:80]: Yep I agree this should be adapted, but it's not really related with this PR Since Mateus is gonna create a dedicated issue for this dynamic content I think you can remove this comment here

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php [hunk 4]
```diff
@@ -354,9 +354,14 @@ public function getListAction(int $productId, ProductCombinationFilters $combina
     public function getCombinationIdsAction(int $productId, ProductCombinationFilters $filters): JsonResponse
     {
         /** @var CombinationRepository $repository */
-        $repository = $this->get('prestashop.adapter.product.combination.repository.combination_repository');
+        $repository = $this->get('PrestaShop\PrestaShop\Adapter\Product\Combination\Repository\CombinationRepository');
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php:359]: I think this modification is not only related to the refacto, if I remember well this API is mostly used by the edition modal so that it has the list of combinations to edit when you use Previous/Next buttons

So this API should be adapted to shop context, but it can be handled separately in one of your PR for combinations multishop because currently there is probably a bug (the modal will display the combinations for all shops even if we are in a single shop with less combinations, so the combination list and the combination modal are not in sync).

TLDR; this is more a bug fix that should be moved in a dedicated PR or one of the PR related to combinations multishop

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

As I told you I think we should delay this PR until the suppliers have been handled for multishop. And I noticed a few modifications which are out of the pure refacto and are actually bug fixes (I commented on a case by case). For those bug fixes they probably deserve to be extracted and fixed early without waiting this PR (especially the combination IDs API).

The idea is to first clean what is not pure refacto in separate PRs so that in the end this PR becomes more simple and only about refactoring

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php (hunk 4 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php:359]: extracted this part here:
https://github.com/PrestaShop/PrestaShop/pull/30585

**Comment by zuk3975** [comment]
closing in favour of https://github.com/PrestaShop/PrestaShop/pull/30797
